File: src/main/java/com/plugin/frege/psi/impl/FregePsiUtilImpl.java
package com.plugin.frege.psi.impl;

import com.intellij.psi.PsiElement;
import com.intellij.psi.util.PsiTreeUtil;
import com.plugin.frege.psi.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class FregePsiUtilImpl {
    private FregePsiUtilImpl() {}

    private static final List<String> defaultImports = List.of(
            "frege.Prelude" // TODO
    );

    private static boolean isScope(@Nullable PsiElement element) {
        return element instanceof FregeScopeElement;
    }

    /**
     * Finds the first parent of the passed element that presents a scope.
     * Returns the passed element, if it is a scope.
     * Returns {@code null} if there is no a scope for the element.
     */
    public static @Nullable FregeScopeElement scopeOfElement(@NotNull PsiElement element) {
        while (element != null && !isScope(element)) {
            element = element.getParent();
        }

        return (FregeScopeElement) element;
    }

    /**
     * Finds a scope of the passed element, gets a list of {@link PsiElement}
     * and applies the passed getter for each of {@link PsiElement}.
     * Also filters {@code nulls} in the resulting list.
     */
    public static @NotNull <T extends PsiElement> List<T> subprogramsFromScopeOfElement(@NotNull PsiElement element,
                                                                                        @NotNull Function<PsiElement, T> getter) {
        FregeScopeElement scope = scopeOfElement(element);
        if (!isScope(scope)) {
            return List.of();
        }

        return scope.getSubprogramsFromScope().stream()
                .map(getter)
                .filter(Objects::nonNull)
                .collect(Collectors.toList());
    }

    /**
     * Finds a scope of the passed element and gets a list of {@link PsiElement}.
     */
    public static @NotNull List<PsiElement> subprogramsFromScopeOfElement(@NotNull PsiElement element) {
        return subprogramsFromScopeOfElement(element, Function.identity());
    }

    /**
     * Returns a predicate, accepting only {@link PsiElement}
     * for which {@link PsiElement#getText()} equals the passed text.
     */
    public static @NotNull Predicate<PsiElement> keepWithText(@NotNull String text) {
        return elem -> elem != null && text.equals(elem.getText());
    }

    /**
     * @return a predicate, accepting only instance of clazz
     * for which {@link PsiElement#getText()} equals the element {@link PsiElement#getText()} if
     * incompleteCode is false.
     */
    public static @NotNull <T extends PsiElement> Predicate<PsiElement> getByTypePredicateCheckingText(
            @NotNull Class<? extends PsiElement> clazz, @NotNull T element, boolean incompleteCode) {
        Predicate<PsiElement> predicate = clazz::isInstance;
        if (!incompleteCode) {
            predicate = predicate.and(elem -> elem != null &&
                    elem.getText().equals(element.getText()));
        }
        return predicate;
    }

    /**
     * Searches for the first binding and if it contains {@link FregeWhereSection}, returns it.
     * Otherwise {@code null} is returned.
     */
    public static @Nullable FregeWhereSection findWhereInExpression(@NotNull PsiElement element) {
        FregeBinding binding = parentBinding(element);
        if (findElementsWithinElementStream(binding, (elem -> elem.equals(element))).noneMatch(x -> true)) {
            return null;
        }
        return (FregeWhereSection) findElementsWithinElementStream(binding, (elem -> elem instanceof FregeWhereSection))
                .findFirst().orElse(null);
    }

    /**
     * Checks if the passed scope is global.
     * @throws IllegalArgumentException if the passed element is not a scope.
     */
    public static boolean isScopeGlobal(@Nullable PsiElement scope) {
        if (!isScope(scope)) {
            throw new IllegalArgumentException("The passed element is not a scope.");
        }

        return scope instanceof FregeBody;
    }

    /**
     * Checks if the passed element is in the global scope.
     * It means that it is one of {@link FregeTopDecl}.
     */
    public static boolean isInGlobalScope(@NotNull PsiElement element) {
        return isScopeGlobal(scopeOfElement(element));
    }

    /**
     * Returns a global scope for the passed element or {@code null} if there is no scope.
     */
    public static @Nullable PsiElement globalScopeOfElement(@Nullable PsiElement element) {
        if (element == null) {
            return null;
        } else if (isScope(element) && isScopeGlobal(element)) {
            return element;
        } else {
            return globalScopeOfElement(element.getParent());
        }
    }

    /**
     * Searches for elements within the scope of the passed element that match the passed predicate.
     */
    public static @NotNull List<PsiElement> findElementsWithinScope(@NotNull PsiElement element,
                                                                    @NotNull Predicate<PsiElement> predicate) {
        List<PsiElement> subprograms = subprogramsFromScopeOfElement(element);
        return subprograms.stream()
                .flatMap(subprogram -> findElementsWithinElementStream(subprogram, predicate))
                .collect(Collectors.toList());
    }

    private static @NotNull Stream<PsiElement> findElementsWithinElementStream(@Nullable PsiElement element,
                                                                               @NotNull Predicate<PsiElement> predicate) {
        if (element == null) {
            return Stream.of();
        }
        if (predicate.test(element)) {
            return Stream.of(element);
        }
        if (isScope(element)) {
            return Stream.of();
        }

        return Arrays.stream(element.getChildren())
                .flatMap(elem -> findElementsWithinElementStream(elem, predicate));
    }

    /**
     * Searches for the elements matched the passed predicate in the children of the passed element within its scope.
     */
    public static @NotNull List<PsiElement> findElementsWithinElement(@Nullable PsiElement element,
                                                                      @NotNull Predicate<PsiElement> predicate) {
        return findElementsWithinElementStream(element, predicate).collect(Collectors.toList());
    }

    /**
     * Returns the first parent which is {@link FregeBinding}.
     */
    public static @Nullable FregeBinding parentBinding(@Nullable PsiElement element) {
        return PsiTreeUtil.getParentOfType(element, FregeBinding.class);
    }

    public static @NotNull List<FregeDataDcl> findAvailableDataDecls(@NotNull PsiElement element) {
        PsiElement globalScope = globalScopeOfElement(element);
        if (globalScope == null) {
            return List.of();
        }
        if (!(globalScope instanceof FregeBody)) {
            throw new IllegalStateException("Global scope must be Frege body.");
        }
        FregeBody body = (FregeBody) globalScope;

        return body.getTopDeclList().stream()
                .map(FregeTopDecl::getDataDcl)
                .filter(Objects::nonNull)
                .collect(Collectors.toList());
    }

    /**
     * @return the module name of psi, if presented, or null otherwise
     */
    public static @Nullable String getModuleName(PsiElement psi) {
        while (psi != null && !(psi instanceof FregeProgram)) {
            psi = psi.getParent();
        }
        if (psi == null)
            throw new IllegalArgumentException("PsiElement is not a part of Frege program");
        FregePackageName packageName = ((FregeProgram) psi).getPackageName();
        if (packageName == null)
            return null;
        return packageName.getText();
    }

    /**
     * Checks if the passed element is a leaf in the PSI tree.
     */
    public static boolean isLeaf(@NotNull PsiElement element) {
        return element.getFirstChild() == null;
    }

    /**
     * Returns all the imports the element can access to.
     * It doesn't return default imports such as 'frege.Prelude'.
     */
    public static @NotNull List<@NotNull FregeImportDcl> findImportsForElement(@NotNull PsiElement element) {
        FregeBody body = PsiTreeUtil.getParentOfType(element, FregeBody.class);
        if (body == null) {
            return List.of();
        }
        return body.getTopDeclList().stream()
                .map(FregeTopDecl::getImportDcl)
                .filter(Objects::nonNull)
                .collect(Collectors.toList());
    }

    /**
     * Returns import names from the file the passed element contains in.
     * If the passed boolean is {@code true}, it includes default imports such as 'frege.Prelude'.
     */
    public static @NotNull List<@NotNull String> findImportsNamesForElement(@NotNull PsiElement element,
                                                                            boolean includingDefault) {
        List<String> imports = FregePsiUtilImpl.findImportsForElement(element).stream()
                .map(FregeImportDcl::getImportPackageName)
                .filter(Objects::nonNull)
                .map(PsiElement::getText)
                .collect(Collectors.toList());
        if (includingDefault) {
            imports.addAll(defaultImports);
        }

        return imports;
    }

    /**
     * Gets the last word after the last '.' in the passed qualified name.
     */
    public static @NotNull String nameFromQualifiedName(@NotNull String qualifiedName) {
        if (!qualifiedName.contains(".")) {
            return qualifiedName;
        }
        return qualifiedName.substring(qualifiedName.lastIndexOf('.') + 1);
    }

    /**
     * Gets the prefix before the last '.' in the passed qualified name.
     */
    public static @NotNull String qualifierFromQualifiedName(@NotNull String qualifiedName) {
        if (!qualifiedName.contains(".")) {
            return "";
        }
        return qualifiedName.substring(0, qualifiedName.lastIndexOf('.'));
    }

    /**
     * Merges full qualified name of class with name (qualified or not) of method or qualified data name.
     * Returns {@code null} if cannot merge.
     * Example: 'frege.prelude.PreludeBase' merges with 'PreludeBase.Int' -> 'frege.prelude.PreludeBase.Int'.
     */
    public static @Nullable String mergeQualifiedNames(@NotNull String first, @NotNull String second) {
        String secondName = nameFromQualifiedName(second);
        String secondQualifier = qualifierFromQualifiedName(second);
        if (secondQualifier.isEmpty() || qualifiedNameEndsWithQualifier(first, secondQualifier)) {
            return first + "." + secondName; // TODO
        } else {
            return null;
        }
    }

    private static boolean qualifiedNameEndsWithQualifier(@NotNull String qualifiedName, @NotNull String qualifier) {
        if (!qualifiedName.endsWith(qualifier)) {
            return false;
        }
        if (qualifiedName.equals(qualifier)) {
            return true;
        }
        return qualifiedName.charAt(qualifiedName.length() - qualifier.length() - 1) == '.';
    }

    /**
     * @return the first parent that inherits FregeDecl, or null if none found
     */
    public static @Nullable PsiElement getDeclType(@NotNull PsiElement element) {
        PsiElement fregeDeclParent = PsiTreeUtil.findFirstParent(element, p -> p instanceof FregeDecl);
        if (fregeDeclParent == null) return null;
        return fregeDeclParent.getFirstChild();
    }
}

File: src/main/kotlin/com/plugin/frege/psi/impl/FregePsiUtilImpl.kt
package com.plugin.frege.psi.impl

import com.intellij.psi.PsiElement
import com.intellij.psi.PsiNamedElement
import com.intellij.psi.tree.TokenSet
import com.intellij.psi.tree.IElementType
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.psi.util.elementType
import com.intellij.psi.util.parentOfType
import com.intellij.psi.util.parentOfTypes
import com.plugin.frege.psi.*
import kotlin.reflect.KClass

object FregePsiUtilImpl {
    fun isScope(element: PsiElement?): Boolean {
        return element is FregeScopeElement
    }

    private fun isWeakScope(element: PsiElement?): Boolean {
        return element is FregeWeakScopeElement
    }

    /**
     * Finds the first parent of [element] that presents a scope.
     * @return the [element], if it is a scope or `null` if there is no a scope for [element].
     */
    @JvmStatic
    fun scopeOfElement(element: PsiElement): FregeScopeElement? {
        return element.parentOfType(true)
    }

    /**
     * The same as [scopeOfElement] but skips scopes which are [FregeWeakScopeElement].
     */
    @JvmStatic
    fun notWeakScopeOfElement(element: PsiElement): FregeScopeElement? {
        var current = element.parentOfType<FregeScopeElement>(true)
        while (current != null) {
            if (!isWeakScope(current)) {
                return current
            }
            current = current.parentOfType(false)
        }
        return null
    }

    /**
     * Finds a scope of [element], gets a list of [PsiElement]
     * and applies [getter] for each of [PsiElement].
     * Also filters `nulls` in the resulting list.
     */
    @JvmStatic
    fun <T : PsiElement> subprogramsFromScopeOfElement(element: PsiElement, getter: (elem: PsiElement) -> T?): List<T> {
        val scope = scopeOfElement(element)
        return if (isScope(scope)) {
            scope!!.subprogramsFromScope.mapNotNull { getter(it) }
        } else {
            emptyList()
        }
    }

    /**
     * @return a scope of [element] and gets a list of subprograms.
     */
    @JvmStatic
    fun subprogramsFromScopeOfElement(element: PsiElement): List<PsiElement> {
        return subprogramsFromScopeOfElement(element) { it }
    }

    /**
     * @return a predicate, accepting only instance of [clazz]
     * for which [PsiNamedElement.getName] equals [name] if [incompleteCode] is `false`
     */
    @JvmStatic
    fun getByTypePredicateCheckingName(
        clazz: KClass<out PsiNamedElement>, name: String, incompleteCode: Boolean
    ): (elem: PsiElement?) -> Boolean {
        val instancePredicate = { elem: PsiElement? -> clazz.isInstance(elem) }
        return if (!incompleteCode) { elem ->
            instancePredicate(elem) && (elem as? PsiNamedElement)?.name == name
        } else {
            instancePredicate
        }
    }

    /**
     * Checks if [scope] is global for a file.
     * @throws IllegalArgumentException if [scope] is not a scope.
     */
    @JvmStatic
    fun isScopeGlobal(scope: PsiElement?): Boolean {
        require(isScope(scope)) { "The passed element is not a scope." }
        return scope is FregeBody
    }

    /**
     * Checks [element] is in the global scope.
     * It means that it is one of [FregeTopDecl].
     */
    @JvmStatic
    fun isInGlobalScope(element: PsiElement): Boolean {
        return isScopeGlobal(notWeakScopeOfElement(element))
    }

    /**
     * @return a global scope for [element] or `null` if there is no scope.
     */
    @JvmStatic
    fun globalScopeOfElement(element: PsiElement?): PsiElement? {
        return if (element == null) {
            null
        } else if (isScope(element) && isScopeGlobal(element)) {
            element
        } else {
            globalScopeOfElement(element.parent)
        }
    }

    /**
     * @return elements within the scope of [element] that match [predicate].
     */
    @JvmStatic
    fun findElementsWithinScope(element: PsiElement, predicate: (elem: PsiElement) -> Boolean): List<PsiElement> {
        val subprograms = subprogramsFromScopeOfElement(element)
        return subprograms.flatMap { findElementsWithinElementSequence(it, predicate) }
    }

    @JvmStatic
    private fun findElementsWithinElementSequence(
        element: PsiElement?,
        predicate: (elem: PsiElement) -> Boolean
    ): Sequence<PsiElement> {
        return when {
            element == null -> {
                emptySequence()
            }
            predicate(element) -> {
                sequenceOf(element)
            }
            isScope(element) && !isWeakScope(element) -> {
                emptySequence()
            }
            else -> {
                element.children.asSequence().flatMap {
                    findElementsWithinElementSequence(it, predicate)
                }
            }
        }
    }

    /**
     * @return elements matched the [predicate] in the children of [element] within its scope.
     */
    @JvmStatic
    fun findElementsWithinElement(
        element: PsiElement?,
        predicate: (elem: PsiElement) -> Boolean
    ): List<PsiElement> {
        return findElementsWithinElementSequence(element, predicate).toList()
    }

    /**
     * @return the module name of [psi], if presented, or `null` otherwise
     */
    @JvmStatic
    fun getModuleName(psi: PsiElement): String? {
        val fregeProgram = psi.parentOfTypes(FregeProgram::class, withSelf = true)
        requireNotNull(fregeProgram) { "PsiElement is not a part of Frege program" }
        return fregeProgram.packageName?.text
    }

    /**
     * Checks if [element] is a leaf in the PSI tree.
     */
    @JvmStatic
    fun isLeaf(element: PsiElement): Boolean {
        return element.firstChild == null
    }

    /**
     * @return the last word after the last '.' in [qualifiedName].
     */
    @JvmStatic
    fun nameFromQualifiedName(qualifiedName: String): String {
        return qualifiedName.substringAfterLast(".")
    }

    /**
     * @return if [name] does contain a qualifier.
     */
    @JvmStatic
    fun isNameQualified(name: String): Boolean {
        return name.contains('.')
    }

    /**
     * Tries to get qualifiers before [usage] and merge them with [usage] text.
     */
    @JvmStatic
    fun getQualifiedNameFromUsage(usage: PsiElement): String {
        val firstQualifier = (usage.prevSibling as? FregeQualifier) ?: return usage.text
        val secondQualifier = (firstQualifier.prevSibling as? FregeQualifier)
            ?: return "${firstQualifier.text}${usage.text}"
        return "${secondQualifier.text}${firstQualifier.text}${usage.text}"
    }

    /**
     * @return the prefix before the last '.' in [qualifiedName].
     */
    @JvmStatic
    fun qualifierFromQualifiedName(qualifiedName: String): String {
        return qualifiedName.substringBeforeLast(".", "")
    }

    /**
     * Merges full qualified name of class with name (qualified or not) of method or qualified data name.
     * Returns `null` if cannot merge.
     * Example: `frege.prelude.PreludeBase` merges with `PreludeBase.Int` -> `frege.prelude.PreludeBase.Int`.
     */
    @JvmStatic
    fun mergeQualifiedNames(first: String, second: String): String {
        val secondName = nameFromQualifiedName(second)
        val secondQualifier = qualifierFromQualifiedName(second)
        return if (qualifiedNameEndsWithQualifier(first, secondQualifier)) {
            "$first.$secondName" // TODO
        } else {
            "$first.$second"
        }
    }

    @JvmStatic
    private fun qualifiedNameEndsWithQualifier(qualifiedName: String, qualifier: String): Boolean {
        return if (!qualifiedName.endsWith(qualifier)) {
            false
        } else if (qualifiedName == qualifier) {
            true
        } else {
            qualifiedName[qualifiedName.length - qualifier.length - 1] == '.'
        }
    }

    /**
     * It is a workaround while we don't have a type system.
     * @return first [FregeConidUsage] from [sigma] if it's not the arrow type. Otherwise `null` will be returned.
     */
    @JvmStatic
    fun findMainTypeFromSigma(sigma: FregeSigma?): FregeConidUsage? {
        val rho = sigma?.rho ?: return null
        if (rho.rho != null) { // it's the arrow type
            return null
        }
        val simpleType = rho.typeApplication?.simpleTypeList?.firstOrNull() ?: return null
        return PsiTreeUtil.findChildOfType(simpleType, FregeConidUsage::class.java)
    }

    /**
     * @return the previous siblings of the PSI element.
     */
    @JvmStatic
    fun siblingBackwardSequenceSkipping(
        element: PsiElement,
        strict: Boolean,
        skip: TokenSet
    ): Sequence<PsiElement> {
        val start = if (strict) element.prevSibling else element
        return generateSequence(start) { it.prevSibling }.filter {
            !skip.contains(it.elementType)
        }
    }

    /**
     * @return is [element] type may separates declarations
     */
    @JvmStatic
    fun isEndDeclElement(element: PsiElement): Boolean {
        val type = element.elementType
        return FregeTypes.VIRTUAL_END_DECL.equals(type) || FregeTypes.SEMICOLON.equals(type)
    }

    /**
     * Checks if within children of [element] there is an PSI node of [type].
     */
    @JvmStatic
    fun isElementTypeWithinChildren(element: PsiElement, type: IElementType): Boolean {
        return generateSequence({ element.firstChild }, { it.nextSibling }).any { it.elementType === type }
    }

    /**
     * Required for navigation to modules from the standard library.
     * According to the language rules, '`frege`' can be omitted in the imported package,
     * provided if you replace the first letter in the next word with the upper-cased one.
     *
     * Example: 'frege.data.HashMap -> Data.HashMap'
     *
     * @return converted [packageName] to the package with the '`frege`' prefix
     * or `null` if it's not possible ([packageName] is empty or the first letter is not upper-cased)
     */
    @JvmStatic
    fun tryConvertToLibraryPackage(packageName: String): String? {
        if (packageName.isEmpty() || !packageName.first().isUpperCase()) {
            return null
        }
        return "frege.${packageName.first().lowercaseChar()}${packageName.substring(1)}"
    }
}
